examples/larger-examples/datalog/subtyping.flix

```diff
@@ -0,0 +1,176 @@
+mod Subtyping {
+
+    type alias Program = {
+        classes = Vector[String],
+        finalClasses = Vector[String],
+        interfaces = Vector[String],
+        implementations = Vector[(String, String)],
+        classExtends = Vector[(String, String)],
+        interfaceExtends = Vector[(String, String)],
+        methods = Vector[(String, String)]
+    }
+
+    def subtypes(program: Program): #{ Subtype(String, String), AvailableMethod(String, String), NonSupertype(String), ExtendsFinalClass(String, String) | r } =
+        let rules = #{
+            Subtype(x, x)       :- Class(x).
+            Subtype(x, x)       :- Interface(x).
+            Subtype(sub, super) :- Class(sub), Interface(super), Implementation(sub, super).
+            Subtype(sub, super) :- Class(sub), Class(super), ClassExtends(sub, super).
+            Subtype(sub, super) :- Interface(sub), Interface(super), InterfaceExtends(sub, super).
+            Subtype(sub, super) :- Subtype(sub, a), Subtype(a, super).
+
+            AvailableMethod(class, method) :- Method(t, method), Subtype(class, t).
+            NonSupertype(class)            :- Class(class), Class(otherClass), not Subtype(otherClass, class), if class != otherClass.
+            ExtendsFinalClass(sub, super)  :- FinalClass(super), Subtype(sub, super), if sub != super.
+        };
+        #(Subtype, AvailableMethod, NonSupertype, ExtendsFinalClass) -> rules <+> toPredicates(program)
+
+    def toPredicates(program: Program): #{ Class(String), FinalClass(String), Interface(String), ClassExtends(String, String),  InterfaceExtends(String, String), Implementation(String, String), Method(String, String) | r } =
+        let cs    = inject program.classes          into Class;
+        let fcs   = inject program.finalClasses     into FinalClass;
+        let is    = inject program.interfaces       into Interface;
+        let impls = inject program.implementations  into Implementation;
+        let ce    = inject program.classExtends     into ClassExtends;
+        let ies   = inject program.interfaceExtends into InterfaceExtends;
+        let ms    = inject program.methods          into Method;
+        cs <+> fcs <+> is <+> ce <+> impls <+> ies <+> ms
+
+    pub def subtypesDistinct(types: Program): Vector[(String, String)] =
+        let v = query subtypes(types) select (x, y) from Subtype(x, y);
+        Vector.filter(match (x, y) -> x != y, v)
+
+    pub def methodsOf(tpe: String, types: Program): Vector[String] =
```

Comment:
I don't want this double layering. Lets just have /one/ function like hasMethod.